app/controllers/devise/two_factor_authentication_controller.rb

```diff
@@ -35,15 +35,13 @@ def update
   private
 
   def check_already_authenticated
-    if user_fully_authenticated? && current_user.unconfirmed_mobile.blank?
-      redirect_to dashboard_index_url
-    end
+    redirect_valid_resource if user_fully_authenticated?
```

Comment:
This is probably what we want, but I wanted to point out the following scenario to see if we agree on the behavior. If they come in via an SP, we set the SAML request url in a session variable, and we keep it around during the session. If they visit `/users/otp` after having authenticated, this `check_already_authenticated` method will make another SAML request, which will take them back to the SP.